metaflow/cmd/diff/__init__.py

```diff
@@ -0,0 +1,198 @@
+            rel = os.path.relpath(dirpath, source_dir)
+            target_file = os.path.join(rel, fname)
+            # source_file is the run file loaded in a tmp directory
+            source_file = os.path.join(dirpath, fname)
+
+            if sys.stdout.isatty() and not output:
+                color = ["--color"]
+            else:
+                color = ["--no-color"]
+
+            if os.path.exists(os.path.join(target_dir, target_file)):
+                cmd = (
+                    ["git", "diff", "--no-index", "--exit-code"]
+                    + color
+                    + [
+                        target_file,
+                        source_file,
+                    ]
+                )
+                result = run(cmd, text=True, stdout=PIPE, cwd=target_dir)
+                if result.returncode == 0:
+                    echo(f"✅ {target_file} is identical, skipping")
+                    continue
+
+                if output:
+                    diffs.append(result.stdout)
+                else:
+                    run(["less", "-R"], input=result.stdout, text=True)
+            else:
+                echo(f"❗ {target_file} not in the target directory, skipping")
+    return diffs if output else None
+
+
+def run_op(runspec, op, op_args):
+    tmp = None
+    try:
+        tmp = extract_code_package(runspec, EXCLUSIONS)
+        op(tmp.name, **op_args)
+    finally:
+        if tmp and os.path.exists(tmp.name):
+            shutil.rmtree(tmp.name)
+
+
+def run_op_diff_runs(source_run, target_run):
+    source_tmp = None
+    target_tmp = None
+    try:
+        source_tmp = extract_code_package(source_run, EXCLUSIONS)
+        target_tmp = extract_code_package(target_run, EXCLUSIONS)
+        perform_diff(source_tmp.name, target_tmp.name)
+    finally:
+        if source_tmp and os.path.exists(source_tmp.name):
+            shutil.rmtree(source_tmp.name)
+        if target_tmp and os.path.exists(target_tmp.name):
+            shutil.rmtree(target_tmp.name)
+
+
+def op_diff(tmpdir):
+    perform_diff(tmpdir)
+
+
+def op_pull(tmpdir, dst=None):
+    if os.path.exists(dst):
+        echo(f"❌  Directory *{dst}* already exists")
+    else:
+        shutil.move(tmpdir, dst)
+        echo(f"Code downloaded to *{dst}*")
+
+
+def op_patch(tmpdir, dst=None):
+    diffs = perform_diff(tmpdir, output=True)
+    with open(dst, "w") as f:
+        for out in diffs:
+            out = out.replace(tmpdir, "/.")
+            out = out.replace("+++ b/./", "+++ b/")
+            out = out.replace("--- b/./", "--- b/")
+            out = out.replace("--- a/./", "--- a/")
+            out = out.replace("+++ a/./", "+++ a/")
+            f.write(out)
+    echo(f"Patch saved in *{dst}*")
+    path = run(
+        ["git", "rev-parse", "--show-prefix"], text=True, stdout=PIPE
+    ).stdout.strip()
+    if path:
+        diropt = f" --directory={path.rstrip('/')}"
+    else:
+        diropt = ""
+    echo("Apply the patch by running:")
```

Comment:
echo and echo_always are a bit confusing here (one calls the other with some foreground, etc). Also, not sure if we want to respect the quiet flag?

Reply:
what is your suggestion? get rid of `echo`?